pydm/tests/widgets/test_spinbox.py

```diff
@@ -0,0 +1,319 @@
+
+    Parameters
+    ----------
+    qtbot : fixture
+        Window for widget testing
+    signals : fixture
+        The signals fixture, which provides access signals to be bound to the appropriate slots
+    monkeypatch : fixture
+        To override default behaviors
+    first_key_pressed : Qt.Key
+        The first key to press to change the spinbox's step exponent value (increase or decrease)
+    second_key_pressed: Qt.Key
+        The second key to press to change the spinbox's step exponent value (increase or decrease)
+    keys_pressed_expected_results : tuple
+        The new value after the first key, and then that after the second key, is pressed
+    """
+    pydm_spinbox = PyDMSpinbox()
+    qtbot.addWidget(pydm_spinbox)
+
+    with qtbot.waitExposed(pydm_spinbox):
+        pydm_spinbox.show()
+
+    pydm_spinbox.step_exponent = 0
+    pydm_spinbox._precision_from_pv = True
+    signals.prec_signal[int].connect(pydm_spinbox.precisionChanged)
+    signals.prec_signal[int].emit(3)
+
+    INIT_SPINBOX_VALUE = 1.2
+    signals.new_value_signal[float].connect(pydm_spinbox.channelValueChanged)
+    signals.new_value_signal[float].emit(INIT_SPINBOX_VALUE)
+
+    pydm_spinbox.setFocus()
+    def wait_focus():
+        return pydm_spinbox.hasFocus()
+    qtbot.waitUntil(wait_focus, timeout=5000)
+
+    def press_key_and_verify(key_pressed, key_mod, key_press_count, expected_exp, expected_value):
+        if key_mod != Qt.NoModifier:
+            # Monkeypatch the Control flag because in this test, we don't properly have the app QApplication instance
+            monkeypatch.setattr(QApplication, "queryKeyboardModifiers", lambda *args: key_mod)
+        else:
+            monkeypatch.setattr(QApplication, "queryKeyboardModifiers", lambda *args: Qt.NoModifier)
+
+        for i in range(0, key_press_count):
+            pydm_spinbox.keyPressEvent(QKeyEvent(QEvent.KeyPress, key_pressed, key_mod))
+        assert pydm_spinbox.step_exponent == expected_exp
+
+        signals.send_value_signal[float].connect(signals.receiveValue)
+        signals.send_value_signal[float].connect(pydm_spinbox.channelValueChanged)
+
+        # Send out the key event to update the value based on the step exponent change
+        pydm_spinbox.keyPressEvent(QKeyEvent(QEvent.KeyPress, Qt.Key_Return, Qt.NoModifier))
+        assert pydm_spinbox.value == expected_value
+
+    # Send out the first key event to effect the change of the step exponent
+    press_key_and_verify(first_key_pressed, Qt.ControlModifier, 2, keys_pressed_expected_results[0],
+                         INIT_SPINBOX_VALUE)
+
+    # Send out the second key press event once to check if it can change the step exponent
+    press_key_and_verify(second_key_pressed, Qt.ControlModifier, 1, keys_pressed_expected_results[1],
+                         INIT_SPINBOX_VALUE)
+
+    # Make sure the widget can process the UpArrow, DownArrow, PageUp, and PageDown keys
```

Comment:
I'm not seeing the `verify` step for those other keys...
I'm also missing the `Return` to send the value and read the proper value at the widget...

Reply:
The up/down arrow and page keys are handled with PyQt default behaviors, so I'm not testing those.

Reply:
Sounds good! If it becomes a problem we can add tests later! 👍